src/Namespace.js

```diff
@@ -0,0 +1,44 @@
+// @flow
+
+import builtinMacros from "./macros";
+import type {MacroMap, MacroExpansion} from "./macros";
+
+export default class Namespace {
+    parent: ?Namespace;
+    macros: MacroMap;
+    //tag: any;
+
+    constructor(parent: ?Namespace, macros: MacroMap = {}) {
+        this.parent = parent;
+        this.macros = macros;
+    }
+
+    getMacro(name: string): ?MacroExpansion {
+        let ns = this;
+        while (ns && !ns.macros.hasOwnProperty(name)) {
+            ns = ns.parent;
+        }
+        if (ns) {
+            return ns.macros[name];
+        } else if (builtinNamespace.macros.hasOwnProperty(name)) {
+            return builtinNamespace.macros[name];
+        } else {
+            return null;
+        }
+    }
+
+    setMacro(name: string, expansion: MacroExpansion, global: Boolean = false) {
+        // Global set is equivalent to setting in all namespaces up the stack.
+        if (global) {
+            let namespace = this;
+            while (namespace) {
+                namespace.macros[name] = expansion;
+                namespace = namespace.parent;
+            }
```

Comment:
What’s the difference between defining the macro at the top-level and defining at all levels?

Reply:
If we only define it at the top level as soon as we exit the current group then it will no longer be defined, or, if it was previously defined using `\def` in that parent group, then the macro would use that definition.

Reply:
I meant the bottom of the stack 😄 But as you said, there is a precedence, so how about deleting the definition if exists and defining only at the bottom?

Reply:
The thing is there could be a definition at any level (or all levels) so `\global\def` has to clobber them all.

Reply:
I was thinking something like:
```js
let namespace = this;
while (namespace.parent) {
    delete namespace.macros[name];
    namespace = namespace.parent;
}
namespace.macros[name] = expansion;
```

Reply:
I see, but then lookup is always expensive for global macros b/c you have to iterate to the root every time.

Reply:
Also, I think `\gdef` should modify `builtinNamespace`.